--- openshift-ansible.orig/playbooks/openstack/inventory.py
+++ openshift-ansible/playbooks/openstack/inventory.py
@@ -81,9 +82,12 @@
 
     # TODO(shadower): filter the servers based on the `OPENSHIFT_CLUSTER`
     # environment variable.
+    no_app_nodes = os.environ.get('INV_NO_APP_NODES', 'false').lower() == "true"
+
     cluster_hosts = [
         server for server in cloud.list_servers()
-        if 'metadata' in server and 'clusterid' in server.metadata]
+        if 'metadata' in server and 'clusterid' in server.metadata and
+            (not no_app_nodes or server.metadata['sub-host-type'] != 'app')]
 
     inventory = base_openshift_inventory(cluster_hosts)
 
