--- openshift-ansible.orig/roles/openshift_openstack/templates/heat_stack.yaml.j2
+++ openshift-ansible/roles/openshift_openstack/templates/heat_stack.yaml.j2
@@ -387,6 +387,11 @@
           protocol: tcp
           port_range_min: 2222
           port_range_max: 2222
+        # iscsi-targets
+        - direction: ingress
+          protocol: tcp
+          port_range_min: 3260
+          port_range_max: 3260
         # heketi dialing backends
         - direction: ingress
           protocol: tcp
@@ -402,6 +407,11 @@
           protocol: tcp
           port_range_min: 24008
           port_range_max: 24008
+        # glusterblockd
+        - direction: ingress
+          protocol: tcp
+          port_range_min: 24010
+          port_range_max: 24010
         # glusterfs_bricks
         - direction: ingress
           protocol: tcp
